pkg/dockerregistry/server/repositorymiddleware.go

```diff
@@ -198,145 +212,41 @@ func (r *repository) Get(dgst digest.Digest) (*schema1.SignedManifest, error) {
-		return nil, err
-	}
-
-	if localImage, err := r.getImage(dgst); err != nil {
-		// if the image is managed by OpenShift and we cannot load the image, report an error
-		if image.Annotations[imageapi.ManagedByOpenShiftAnnotation] == "true" {
-			context.GetLogger(r.ctx).Errorf("Error getting image %q: %v", dgst.String(), err)
-			return nil, err
-		}
-	} else {
-		// if we have a local manifest, use it
-		if len(localImage.DockerImageManifest) > 0 {
-			return r.manifestFromImageWithCachedLayers(localImage, cacheName)
-		}
-	}
+	manifest, err := r.manifestFromImageWithCachedLayers(image, ref.DockerClientDefaults().Exact(), supportsSchema2)
 
-	// allow pullthrough to be disabled
-	if !r.pullthrough {
-		return nil, distribution.ErrManifestBlobUnknown{Digest: dgst}
-	}
-
-	// check the previous error here
-	if referenceErr != nil {
-		context.GetLogger(r.ctx).Errorf("Error parsing image %q: %v", image.DockerImageReference, referenceErr)
-		return nil, referenceErr
-	}
-
-	return r.pullthroughGetByTag(image, ref, cacheName, options...)
+	return manifest, err
 }
 
-// pullthroughGetByTag attempts to load the given image manifest from the remote server defined by ref, using cacheName to store any cached layers.
-func (r *repository) pullthroughGetByTag(image *imageapi.Image, ref imageapi.DockerImageReference, cacheName string, options ...distribution.ManifestServiceOption) (*schema1.SignedManifest, error) {
-	defaultRef := ref.DockerClientDefaults()
-
-	retriever := r.importContext()
-
-	repo, err := retriever.Repository(r.ctx, defaultRef.RegistryURL(), defaultRef.RepositoryName(), false)
-	if err != nil {
-		context.GetLogger(r.ctx).Errorf("Error getting remote repository for image %q: %v", image.DockerImageReference, err)
-		return nil, err
-	}
-
-	// get a manifest context
-	manifests, err := repo.Manifests(r.ctx)
-	if err != nil {
-		context.GetLogger(r.ctx).Errorf("Error getting manifests for image %q: %v", image.DockerImageReference, err)
-		return nil, err
+// Put creates or updates the named manifest.
+func (r *repository) Put(ctx context.Context, manifest distribution.Manifest, options ...distribution.ManifestServiceOption) (digest.Digest, error) {
+	switch manifest.(type) {
+	case *schema1.SignedManifest:
+	case *schema2.DeserializedManifest:
+	default:
+		return "", fmt.Errorf("unrecognized manifest type %T", manifest)
 	}
 
-	// fetch this by image
-	if len(ref.ID) > 0 {
-		dgst, err := digest.ParseDigest(ref.ID)
-		if err != nil {
-			context.GetLogger(r.ctx).Errorf("Error getting manifests for image %q: %v", image.DockerImageReference, err)
-			return nil, err
-		}
-		manifest, err := manifests.Get(dgst)
-		if err != nil {
-			context.GetLogger(r.ctx).Errorf("Error getting manifest from remote server for image %q: %v", image.DockerImageReference, err)
-			return nil, err
+	if !r.allowSchema2 {
+		if _, ok := manifest.(*schema1.SignedManifest); !ok {
+			return "", fmt.Errorf("schema version 2 disabled")
```

Comment:
What does registry 2.2 return for schema2 manifest? We should return the same HTTP code.